### PR TITLE
Android - implement ChipDnssdStopBrowse

### PR DESCRIPTION
**Pull request by pgregorr-amazon**
This change adds the implementation for ChipDnssdStopBrowse in connectedhomeip/src/platform/android/DnssdImpl.cpp allowing DNS-SD discovery to be manally stopped before the timeout.

**Change summary**

1.	Implemented the ChipDnssdStopBrowse JNI function in connectedhomeip/src/platform/android/DnssdImpl.cpp which calls the NsdManagerServiceBrowser stopDiscover() Java method (in connectedhomeip/src/platform/android/java/chip/platform/NsdManagerServiceBrowser.java) This function stops ongoing browsing for services using DNS-SD.

**Testing**

Verified and tested locally with the Android tv-casting-app example app. By calling the stop discovery API form the tv-casting-app app GUI, we are able to stop NsdManagerDiscovery discovery, prior to timeout triggered stop discovery.

### CODE REVIEW

File: examples/tv-casting-app/android/App/app/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -31,8 +31,12 @@
     <string name="matter_stop_discovery_button_text">Stop Discovery"</string>
     <string name="matter_clear_results_button_text">Clear Results"</string>
     <string name="matter_discovery_message_initializing_text">Initializing</string>
-    <string name="matter_discovery_message_discovering_text">Casting Players on-network:</string>
+    <string name="matter_discovery_message_discovering_text">Discovering Casting Players on-network:</string>
     <string name="matter_discovery_message_stopped_text">Discovery Stopped</string>
+    <string name="matter_discovery_error_message_initial">No errors to report</string>
+    <string name="matter_discovery_error_message_start_error">START DISCOVERY error. -</string>
+    <string name="matter_discovery_error_message_stop_before_starting">START DISCOVERY error. Discovery ongoing, STOP before STARTING. -</string>
+    <string name="matter_discovery_error_message_stop_error">STOP DISCOVERY error. -</string>
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/android/App/app/src/main/res/values/strings.xml:39]: Lower case is ok

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayerDiscovery-JNI.cpp [hunk 2]
```diff
@@ -246,6 +246,13 @@ JNI_METHOD(jobject, removeCastingPlayerChangeListener)(JNIEnv * env, jobject, jo
 
         return support::convertMatterErrorFromCppToJava(CHIP_NO_ERROR);
     }
+    else if (DiscoveryDelegateImpl::GetInstance()->castingPlayerChangeListenerJavaObject.ObjectRef() == nullptr)
+    {
+        ChipLogError(AppServer,
+                     "CastingPlayerDiscovery-JNI::removeCastingPlayerChangeListener() Warning: Cannot remove listener. No "
```
Comment (sharadb-amazon) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayerDiscovery-JNI.cpp:252]: Please remove the word "Warning" when calling ChipLogError. "Error" and "Warning" are usually different log levels.

**Review by sharadb-amazon — APPROVED** [review]

File: src/lib/dnssd/platform/Dnssd.h [hunk 3]
```diff
@@ -230,8 +230,10 @@ CHIP_ERROR ChipDnssdBrowse(const char * type, DnssdServiceProtocol protocol, chi
  *
  * @param browseIdentifier an identifier for a browse operation that came from
  *                         ChipDnssdBrowse.
+ * @param callback         The callback to call when the browse is stopped.
+ *
  */
-CHIP_ERROR ChipDnssdStopBrowse(intptr_t browseIdentifier);
+CHIP_ERROR ChipDnssdStopBrowse(intptr_t browseIdentifier, DnssdBrowseCallback callback = {});
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/platform/Dnssd.h:236]: This doesn't make sense.   The API documentation is pretty clear: the DnssdBrowseCallback that gets called is the one that was passed to ChipDnssdBrowse.  What is the point of this second DnssdBrowseCallback?

Also note that doing this break all the other platforms...

**Review by bzbarsky-apple — DISMISSED** [review]

File: src/lib/dnssd/platform/Dnssd.h (hunk 3 above)
Comment (pgregorr-amazon) [reply src/lib/dnssd/platform/Dnssd.h]: Right. This is causing CI failures in the other platforms... I thought it would be treated as an optional parameter. 
connectedhomeip/src/platform/android/java/chip/platform/NsdManagerServiceBrowser.java  stopDiscover() was expecting this second DnssdBrowseCallback but it can be retained in stopDiscover() instead of being passed around. Addressed in latest commit.

File: examples/tv-casting-app/android/App/app/src/main/res/values/strings.xml (hunk 1 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/android/App/app/src/main/res/values/strings.xml]: Updated.

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayerDiscovery-JNI.cpp (hunk 2 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayerDiscovery-JNI.cpp]: Removed.

File: src/lib/dnssd/platform/Dnssd.h [hunk 4]
```diff
@@ -230,6 +230,7 @@ CHIP_ERROR ChipDnssdBrowse(const char * type, DnssdServiceProtocol protocol, chi
  *
  * @param browseIdentifier an identifier for a browse operation that came from
  *                         ChipDnssdBrowse.
+ *
```
Comment (bzbarsky-apple) [inline src/lib/dnssd/platform/Dnssd.h:233]: Why this change?

File: src/platform/android/DnssdImpl.cpp [hunk 5]
```diff
@@ -209,7 +210,20 @@ CHIP_ERROR ChipDnssdBrowse(const char * type, DnssdServiceProtocol protocol, Ine
 
 CHIP_ERROR ChipDnssdStopBrowse(intptr_t browseIdentifier)
 {
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    VerifyOrReturnError(sMdnsCallbackObject.HasValidObjectRef(), CHIP_ERROR_INCORRECT_STATE);
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+
+    env->CallVoidMethod(sBrowserObject.ObjectRef(), sStopBrowseMethod, sMdnsCallbackObject.ObjectRef());
```
Comment (bzbarsky-apple) [inline src/platform/android/DnssdImpl.cpp:217]: This is not doing anything with browseIdentifier, right?

So if the core code starts an operational browse, then a commissionable browse, then stops the commissionable browse, what will the behavior here be?

File: src/lib/dnssd/platform/Dnssd.h (hunk 4 above)
Comment (pgregorr-amazon) [reply src/lib/dnssd/platform/Dnssd.h]: Sorry, missed this extra line. Removed.

File: src/platform/android/DnssdImpl.cpp (hunk 5 above)
Comment (pgregorr-amazon) [reply src/platform/android/DnssdImpl.cpp]: Right. In the latest commit, we're now utilizing browseIdentifier to stop DNS-SD browse, similar to Darwin.
For the second question.  Multiple browse operations are currently not possible on Android. We plan to implement that in the future.

File: src/platform/android/java/chip/platform/NsdManagerServiceBrowser.java [hunk 6]
```diff
@@ -81,23 +89,30 @@ public void startDiscover(
       final long contextHandle,
       final ChipMdnsCallback chipMdnsCallback) {
     if (callbackMap.containsKey(callbackHandle)) {
-      Log.d(TAG, "Invalid callbackHandle");
+      Log.w(TAG, "Starting service discovering failed. Invalid callback ID: " + callbackHandle);
```
Comment (sharadb-amazon) [inline src/platform/android/java/chip/platform/NsdManagerServiceBrowser.java:92]: Just call it "callbackHandle" instead of "callback ID" in the string because ID could mean something else and is confusing to see in this context.

**Review by joonhaengHeo — APPROVED** [review]

File: src/platform/android/DnssdImpl.cpp [hunk 7]
```diff
@@ -203,13 +204,31 @@ CHIP_ERROR ChipDnssdBrowse(const char * type, DnssdServiceProtocol protocol, Ine
...
+    auto ctx     = reinterpret_cast<BrowseContext *>(browseIdentifier);
+
+    env->CallVoidMethod(sBrowserObject.ObjectRef(), sStopBrowseMethod, reinterpret_cast<jlong>(ctx->callback));
+
+    if (env->ExceptionCheck())
+    {
+        ChipLogError(Discovery, "Java exception in ChipDnssdStopBrowse");
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+        return CHIP_JNI_ERROR_EXCEPTION_THROWN;
```
Comment (bzbarsky-apple) [inline src/platform/android/DnssdImpl.cpp:227]: You just leaked `ctx`.

Comment (bzbarsky-apple) [reply src/platform/android/DnssdImpl.cpp:227]: @pgregorr-amazon See above^

Comment (pgregorr-amazon) [reply src/platform/android/DnssdImpl.cpp:227]: Ahh. Thanks for catching that! Submitted a PR with fix for that: https://github.com/project-chip/connectedhomeip/pull/32855

File: src/platform/android/DnssdImpl.cpp [hunk 8]
```diff
@@ -203,13 +204,31 @@ CHIP_ERROR ChipDnssdBrowse(const char * type, DnssdServiceProtocol protocol, Ine
...
+
     return CHIP_NO_ERROR;
 }
 
 CHIP_ERROR ChipDnssdStopBrowse(intptr_t browseIdentifier)
 {
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    VerifyOrReturnError(sBrowserObject.HasValidObjectRef() && sStopBrowseMethod != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
```
Comment (yunhanw-google) [inline src/platform/android/DnssdImpl.cpp:217]: please check whether env is null or not and print error.

File: src/platform/android/DnssdImpl.cpp [hunk 9]
```diff
@@ -203,13 +204,31 @@ CHIP_ERROR ChipDnssdBrowse(const char * type, DnssdServiceProtocol protocol, Ine
...
+    auto ctx     = reinterpret_cast<BrowseContext *>(browseIdentifier);
```
Comment (yunhanw-google) [inline src/platform/android/DnssdImpl.cpp:218]: please also check whether browseIdentifier is null or not

File: src/platform/android/DnssdImpl.cpp [hunk 10]
```diff
@@ -203,13 +204,31 @@ CHIP_ERROR ChipDnssdBrowse(const char * type, DnssdServiceProtocol protocol, Ine
...
+
+    if (env->ExceptionCheck())
+    {
+        ChipLogError(Discovery, "Java exception in ChipDnssdStopBrowse");
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+        return CHIP_JNI_ERROR_EXCEPTION_THROWN;
+    }
+    chip::Platform::Delete(ctx);
+
```
Comment (yunhanw-google) [inline src/platform/android/DnssdImpl.cpp:230]: usually we set ctx as null for style.

File: src/platform/android/java/chip/platform/NsdManagerServiceBrowser.java [hunk 11]
```diff
@@ -81,23 +89,30 @@ public void startDiscover(
       final long contextHandle,
       final ChipMdnsCallback chipMdnsCallback) {
     if (callbackMap.containsKey(callbackHandle)) {
-      Log.d(TAG, "Invalid callbackHandle");
+      Log.w(TAG, "Starting service discovery failed. Invalid callbackHandle: " + callbackHandle);
       return;
     }
 
     NsdManagerDiscovery discovery =
```
Comment (yunhanw-google) [inline src/platform/android/java/chip/platform/NsdManagerServiceBrowser.java:96]: please also check whether this new is allocating memory successfully?

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayerDiscovery-JNI.cpp [hunk 12]
```diff
@@ -246,11 +246,18 @@ JNI_METHOD(jobject, removeCastingPlayerChangeListener)(JNIEnv * env, jobject, jo
 
         return support::convertMatterErrorFromCppToJava(CHIP_NO_ERROR);
     }
+    else if (DiscoveryDelegateImpl::GetInstance()->castingPlayerChangeListenerJavaObject.ObjectRef() == nullptr)
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayerDiscovery-JNI.cpp:249]: please use HasValidObjectRef

File: src/platform/android/DnssdImpl.cpp [hunk 13]
```diff
@@ -306,6 +325,8 @@ void InitializeWithObjects(jobject resolverObject, jobject browserObject, jobjec
 
     sBrowseMethod = env->GetMethodID(browserClass, "browse", "(Ljava/lang/String;JJLchip/platform/ChipMdnsCallback;)V");
 
+    sStopBrowseMethod = env->GetMethodID(browserClass, "stopDiscover", "(J)V");
```
Comment (yunhanw-google) [inline src/platform/android/DnssdImpl.cpp:328]: you also need check whether sStopBrowseMethod == nullptr? right? similar to sResolveMethod

**Review by yunhanw-google — COMMENTED** [review]
please address the comments in follow-up, thanks

File: src/platform/android/DnssdImpl.cpp (hunk 13 above)
Comment (pgregorr-amazon) [reply src/platform/android/DnssdImpl.cpp:328]: Right

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayerDiscovery-JNI.cpp (hunk 12 above)
Comment (pgregorr-amazon) [reply examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayerDiscovery-JNI.cpp:249]: Okay

File: src/platform/android/DnssdImpl.cpp (hunk 9 above)
Comment (pgregorr-amazon) [reply src/platform/android/DnssdImpl.cpp:218]: Okay

File: src/platform/android/DnssdImpl.cpp (hunk 10 above)
Comment (pgregorr-amazon) [reply src/platform/android/DnssdImpl.cpp:230]: Adding ctx = nullptr;
